rfcs/NNNN-preact-signals.md

```diff
@@ -0,0 +1,184 @@
+          // This will cause another call into performUpdate, which will
+          // then create a new effect watching that update pass.
+          this.requestUpdate();
+        }
+        updateFromLit = false;
+      });
+    }
+```
+
+### watch() directive
+
+The `watch()` async directive accepts a signal and renders its value synchronously to the containing binding. When the signal changes, the binding value is updated directly.
+
+Usage:
+```ts
+html`<p>${watch(messageSignal)}</p>`
+```
+
+```ts
+class WatchSignal extends AsyncDirective {
+  render(signal: Signal<unknown>) {
+    let updateFromLit = true;
+    effect(() => {
+      // Access the signal unconditionally to watch it
+      const value = signal.value;
+      if (!updateFromLit) {
+        this.setValue(value);
+      }
+      updateFromLit = false;
+    });
+    return signal.value;
+  }
+}
+export const resolvePromise = directive(ResolvePromise);
+```
+
+### auto-watching template tag
+
+Auto-watching versions of `html` and `svg` template tags will scan a template result's values and automatically wrap them in a `watch()` directive if they are signals.
+
+We should be able to detect signals with `value instanceof Signal`. Which the `instanceof` operator is fragile, especially in the presence of multiple copies of a module, it appears that the Preact Signals modules are _already_ fragile in this respect due to module-scoped state: all signals, computed signals, and effects, must use the same module instance to work.
+
+## Implementation Considerations
+
+### Implementation Plan
+
+TBD
+
+### Backward Compatibility
+
+N/A
+
+### Testing Plan
+
+Unit tests are sufficient for client-side rendering. We should also include server-side tests with the SSR fixture utility.
+
+### Performance and Code Size Impact
+
+No impact on core library size or performance.
+
+### Interoperability
+
+Signals implementations are unfortunately not interoperable with each other. Signals are most appropriate for the internal state of a component, or share state amongst already tightly-coupled components, as in an app.
+
+### Security Impact
+
+N/A
+
+### Documentation Plan
+
+This package will initially be documented in it's own README. If it stays on track to graduation, we should document this package under the *Managing Data* section on lit.dev. We may end up with multiple signals packages, and either none will graduate, one will, or we'll have to document multiple packages.
+
+## Downsides
+
+Implementing this RFC may appear as endorsing Preact Signals above other signals packages. This will be true to the extent that we will have built and tested the integration ourselves, so know it works. The downside is that other signals implementations may be as good or better choices, and our intention of Preact Signals integration only being the first integration could be overlooked causing people to not use the other packages. Something has to go first however.
```

Comment:
Are there any downsides to creating a generalized solution where `effect` and `Signal` can be injected or configured instead of being hard coded?  That way the solution can be generalized from the start and the issue of selecting Preact Signals is left as an implementation detail.

Reply:
Each signal library has it's own API. I'm not sure that an abstraction we designed now would fit every one. Plus, Overriding `performUpdate()` basically _is_ that abstraction. Each signal library can have it's own mixin and an element could apply multiple signal watchers if you wanted.

Reply:
That makes sense. It probably is not worth the extra effort to create such an abstraction considering how small this actually is. Not even taking into account the issue of getting the abstractuon right to beging with. Thanks for the reply.